baybe/searchspace/discrete.py

```diff
@@ -697,33 +728,40 @@ def _apply_polars_constraint_filter(
     return ldf
 
 
-def _apply_constraint_filter(
-    df: pd.DataFrame, constraints: Collection[DiscreteConstraint]
-):
-    """Remove discrete search space entries inplace based on constraints.
+def parameter_cartesian_prod_polars(parameters: Iterable[Parameter]) -> pl.LazyFrame:
+    """Create the Cartesian product of all parameter values.
+
+    Ignores continuous parameters.
 
     Args:
-        df: The data in experimental representation to be modified inplace.
-        constraints: List of discrete constraints.
+        parameters: List of parameter objects.
 
+    Returns:
+        A lazy dataframe containing all possible discrete parameter value combinations.
     """
-    # Reorder the constraints according to their execution order
-    constraints = sorted(
-        constraints,
-        key=lambda x: DISCRETE_CONSTRAINTS_FILTERING_ORDER.index(x.__class__),
-    )
+    discrete_parameters = [p for p in parameters if p.is_discrete]
+    if not discrete_parameters:
+        return pl.LazyFrame()
 
-    # Remove entries that violate parameter constraints:
-    for constraint in (c for c in constraints if c.eval_during_creation):
-        idxs = constraint.get_invalid(df)
-        df.drop(index=idxs, inplace=True)
-    df.reset_index(inplace=True, drop=True)
+    # Convert each parameter to a lazy dataframe for cross-join operation
+    param_frames = [pl.LazyFrame({p.name: p.values}) for p in discrete_parameters]  # type:ignore[attr-defined]
+
+    # Handling edge cases
+    if len(param_frames) == 1:
+        return param_frames[0]
 
+    # Cross-join parameters
+    res = param_frames[0]
+    for frame in param_frames[1:]:
+        res = res.join(frame, how="cross", force_parallel=True)
```

Comment:
I'm still not very clear about this `force_parallel` option, especially since there is also an `allow_parallel` option. Can we again summarize what is the intention here and why not omit it / use the other option? In the end, I think the code at least deserves a comment, but perhaps let's first clarify here

Reply:
Using `allow_parallel` doesn't guarantee parallelization. 
The optimizer may still choose not to parallelize if it deems it unnecessary or potentially detrimental. `force_parallel`  overrides the optimizer's decision and ensures parallel execution

Reply:
it seems to me that `allow_parallel` is the way to go then, is there any reason why you opted for `force_parallel` @rjavadi  ?

Reply:
I'd also prefer to go with the least invasive option (since we are not sure about potential side effects)

Reply:
I explained in the last [here](https://github.com/emdgroup/baybe/discussions/267) why I used `force_parallel`.

Reply:
hmm comparing the risks here, it seems `force_parallel` can do speedup for large spaces. Given the prelim investigation it also seems that it was necessary to set this flag as without it polars did not opt for the parallelization. `allow_parallel` on the other hand seems the pick for small spaces, as then the parallelization is turned off if detrimental. But since the small space regime doesnt really matter (even if there was a difference of 10x computation time depending on which flag to choose) then I would leave `force_parallel`, @AdrianSosic lets do that in the interest of moving forward we can change it anytime again if issues would be encountered

Reply:
Alright 👍🏼